Preprocessing profiles

jameshadfield opened the pull request:
This breaks each of our main nextstrain profiles into two profiles each - one to perform alignment (etc) and a second to run the phylogenetics. This is part of a wider effort to generate alignments as soon as new sequences are available.

**First commit**

Using GISAID as an example (also implemented for open, read the following but replace `gisaid` with `open` ):

The first profile `nextstrain_profiles/nextstrain-gisaid-preprocess` takes sequences and metadata and produces four files which we upload to S3,  `results/filtered_gisaid.fasta.xz`, `results/masked_gisaid.fasta.xz`, `mutation-summary.tsv.xz` and `results/aligned_gisaid.fasta.xz`.  While this is a separate profile, the  `builds.yaml` file is around a dozen lines of code.

The second profile, `nextstrain_profiles/nextstrain-gisaid` is relatively unchanged, except that we now start from `results/filtered_gisaid.fasta.xz` and thus should be much faster to run. Note that the `upload`  rule here will no longer upload the files which are now within the previous profile. Iit is still possible to start this workflow from (unaligned) sequences, however there should be no reason to do so.

Note that both profiles will run `rule sanitize_metadata`, as each depend on its output which is not uploaded as an intermediate file.

None of these changes should affect non-nextstrain-core builds / profiles.

**Second commit**

A GitHub action to allow the preprocessing profiles to be triggered once the relevant nextstrain/ncov-ingest workflows finish via a repository dispatch event. ~This commit is a WIP~


**Next steps**

- [x] Test preprocess profile on AWS - [In progress](https://console.aws.amazon.com/batch/home?region=us-east-1#jobs/detail/d04b02fa-0c95-4cd1-8ad9-ddff61e867fa)
- [x] Test (phylogenetic) build profiles - [In progress](https://console.aws.amazon.com/batch/home?region=us-east-1#jobs/detail/35d7d2c5-c495-4384-9374-6e95eedeab85)
- [x] Test this GitHub action (you should be able to use a POST request, [even on a branch](https://docs.github.com/en/rest/reference/repos#create-a-repository-dispatch-event), I think). _UPDATE: tested using a push event, which will now be removed_
- [x] Modify actions to upload to the correct S3 address & add a GISAID action
- [x] Test a range of nextalign runs with different numbers of threads to see how long this will take - _This is currently in progress_
- [x] Ensure alignment is comfortably within the 6h run time limit, else we need `--dispatch` _UPDATE: using --detach_
- [ ] Merge this PR
- [ ] Add the workflow dispatch REST calls to [fetch-and-ingest-genbank-master](https://github.com/nextstrain/ncov-ingest/blob/master/.github/workflows/fetch-and-ingest-genbank-master.yml) and [fetch-and-ingest-gisaid-master](https://github.com/nextstrain/ncov-ingest/blob/master/.github/workflows/fetch-and-ingest-gisaid-master.yml)
- [ ] (bonus) if the resulting builds which start from `filtered.fasta.xz` are comfortably within 6 hours, we can remove `--detach` from any actions which run those profiles. _UPDATE they take less than 3 hours_
- [ ] ??? Create rebuild actions in this repo which are either triggered via workflow dispatches or on a schedule

huddlej reviewed (commented):
Although I don't know much about GH Actions or ncov-ingest, this looks really good. I'm impressed by how cleanly you're able to "split" the workflow by just calling different end-points and how other users of the workflow won't be affected by these changes!

As I read through these changes and your PR description, I wondered whether we should include some of these details about how and why the profiles are structured in the `docs/dev_docs.md`. Does that seem like an appropriate place?

eharkins reviewed (commented):
> As I read through these changes and your PR description, I wondered whether we should include some of these details about how and why the profiles are structured in the docs/dev_docs.md. Does that seem like an appropriate place?

Does this PR also warrant a release / update to https://github.com/nextstrain/ncov/blob/master/docs/src/reference/change_log.md or perhaps not if it doesn't introduce breaking changes?


> Test this GitHub action (you should be able to use a POST request, even on a branch, I think). UPDATE: tested using a push event, which will now be removed

I only see `preprocess-open` on https://github.com/nextstrain/ncov/actions. Did you also test the `gisaid` counterpart?

>  Add the workflow dispatch REST calls to fetch-and-ingest-genbank-master and fetch-and-ingest-gisaid-master

this == https://github.com/nextstrain/ncov-ingest/pull/212, correct?

>  ??? Create rebuild actions in this repo which are either triggered via workflow dispatches or on a schedule

If we migrate the code:
- https://github.com/nextstrain/ncov-ingest/blob/master/.github/workflows/rebuild-gisaid.yml
- https://github.com/nextstrain/ncov-ingest/blob/master/bin/rebuild
- https://github.com/nextstrain/ncov-ingest/blob/master/bin/launch-build

which accomplishes this from the `ncov-ingest` repo to this repo and adjust it to run in the context of this repo, do we need to change the arguments passed to the `launch-build` script or will it pick up from the alignment thanks to your changes here.

One thing that happens when we run a rebuild from an action in `ncov-ingest` is slack notifications. We could still get those for launching rebuilds if we use that same code as I suggested above. Do we want similar notifications for when preprocess is run here?

tsibley reviewed (commented):
Thanks for tackling this, @jameshadfield! I think its on the right track.

However, there's a lot of duplication in these changes, and I think it would be better to avoid more of it where we can. We've already hit divergence issues just with the duplication between existing profiles, and more divergence feels inevitable if we duplicate more profiles and rules.

For example, I think there must be a better way to factor the profiles/build configs so we can use the existing `nextstrain-open` and `nextstrain-gisaid` profiles to do both "standard" builds (starting from "filtered") and "preprocess" jobs (starting from "sequences") without having to create new stubby profiles.  I think one way, for instance, would be to provide an extra `builds-full.yaml` which just contained

    inputs:
      - name: open
        metadata: s3://…
        sequences: s3://…

and then run "preprocess" jobs with it as `--profile nextstrain_profiles/nextstrain-open --configfile nextstrain_profiles/nextstrain-open/builds-preprocess.yaml`.

Similarly, the new file `export_preprocessed_for_nextstrain.smk` which overrides existing rules in the workflow feels like a fragile and brittle way to accomplish the behaviour changes needed.  For example, `_get_upload_inputs()` could conditionalize what it returns based on the build inputs configured rather than become an additional `_get_preprocessed_upload_inputs()` which is called by way of redefining the `upload` rule in `export_preprocessed_for_nextstrain.smk`.  The onstart/onerror handlers could be similarly updated.

I also think its worth considering if "preprocess" is the right name here.  I know it has a long storied past in Nextstrain history ;-), but "process" is a pretty generic term like "data".  A more precise word might serve us better in the long run, especially if we can come up with complementary way to describe the "standard" or "from filtered" builds at the same time.

Finally, I think it will ease development and testing (and avoid accidental overwrites of production data) if the two new "preprocess" GH Actions workflows supported branch-specific output (and maybe even input?) locations, like the existing ncov-ingest jobs (input + output) and ncov trial jobs (output only).

jameshadfield commented:
Thanks @trs. Duplication is a valid concern, and one I'd like to address. I can switch to a single profile with two configfiles, and invoke the preprocessing with an additional `--configfile` argument, as opposed to two profiles. This is a little less duplication at the cost of a little more complexity (configs are now overlaid). 

> Similarly, the new file `export_preprocessed_for_nextstrain.smk` which overrides existing rules in the workflow feels like a fragile and brittle way to accomplish the behaviour changes needed. 

It doesn't override them. The normal profile imports the custom rule `workflow/snakemake_rules/export_for_nextstrain.smk` while the preprocess profile imports `workflow/snakemake_rules/export_preprocessed_for_nextstrain.smk`. I think having each profile (or configfile if we go that way) have a specific `upload` rule is simple and easy to remember. Within those rules, the only duplicated code is the `onstart / onerror` handlers; I will look into deduplicating these. 

> I think it will ease development and testing (and avoid accidental overwrites of production data) if the two new "preprocess" GH Actions workflows supported branch-specific output (and maybe even input?) locations, like the existing ncov-ingest jobs (input + output) and ncov trial jobs (output only).

The preprocessing sees less code changes compared with ingest and the normal builds, so I was hoping we could avoid adding more actions here and, if needed, deploy builds from our own machine. However I can for sure follow `ncov-ingest` here and make these extra actions.

tsibley commented:
> It doesn't override them. … I think having each profile (or configfile if we go that way) have a specific `upload` rule is simple and easy to remember.

Nod. I see that it doesn't override them _at runtime_, but it does _conceptually_. That is, there are two different implementations of the `upload` rule. In my experience, that often impedes maintenance later.

> The preprocessing sees less code changes compared with ingest and the normal builds, so I was hoping we could avoid adding more actions here and, if needed, deploy builds from our own machine. However I can for sure follow `ncov-ingest` here and make these extra actions.

Ah, I don't think there need to be additional separate GH Actions workflows, just parameterizing what you already have. For example, something similar to one of these two parameterizations on branch/ref: [example 1](https://github.com/nextstrain/ncov-ingest/blob/5db5d2574210d60b4ae6434248af64dd3187a781/bin/ingest-genbank#L41-L61) and [example 2](https://github.com/nextstrain/ncov/blob/71508e0a31935c0682f2ad4b17ce862f2328c229/.github/workflows/trial-build.yml#L40).

jameshadfield commented:
I have chosen to continue with “stubby profiles” as I think this is the easiest to understand solution. Using `--configfile` ends up _replacing_ any `configfile` entries defined in the profile’s `config.yaml`, so the additional `builds_preprocess.yaml` ends up looking like the one in a new profile anyway. In general, how Snakemake overlays various configs isn’t intuitive (for instance, [see these comments we made to our future selves](https://github.com/nextstrain/ncov/blob/master/Snakefile#L19-L38)), and I prefer a more verbose but easier to reason with implementation. 

I have, however, removed the duplicate `upload` rules following review feedback; we now parameterise the `_get_upload_inputs` function instead. This results in three possible behaviours of the `upload` rule:
	1. You set `upload_preprocessing_files` in the config. This does what it says :)
	2. The config starts from the filtered stage. This uploads the phylogenetic build files.
	3. For any other config setup, it’s unclear which files we wish to upload, so we prefer to exit loudly rather than perform an ambiguous action.

> As I read through these changes and your PR description, I wondered whether we should include some of these details about how and why the profiles are structured in the docs/dev_docs.md. Does that seem like an appropriate place?

Great idea. I’ve added docs there. Now the task is to keep them up to date as other PRs change the way we automate these pipelines!

> Finally, I think it will ease development and testing (and avoid accidental overwrites of production data) if the two new “preprocess” GH Actions workflows supported branch-specific output

Done… I think! It uses [the “fake” ternary available to GitHub expressions](https://github.community/t/do-expressions-support-ternary-operators-to-change-their-returned-value/18114/3) so that we can use the same actions for trial builds, thus avoiding duplication and the inevitable divergence. Would be great to actually talk through this with someone, I know many people here are more familiar with GitHub actions than I am.

jameshadfield commented:
I've tested this GitHub Action in a variety of ways [using this testing repo](https://github.com/nextstrain/test-github-actions/actions/workflows/preprocess-open.yml) as well as testing the underlying snakemake scripts via AWS.

eharkins commented:
> The launch-build script will continue to run the (e.g.) nextstrain_profiles/nextstrain-open profile, and the only change introduced here is that this will now start from the filtered alignment; no changes to arguments are needed.

To clarify: is the plan to continue with our manual triggering (e.g. https://github.com/nextstrain/ncov-ingest/blob/master/.github/workflows/rebuild-gisaid.yml) of the phylogeny-building part of the workflow, and then we can choose to do either or both of the following in subsequent PRs? :
- migrating those workflows to this repository
- setting them to run on a schedule or triggering them directly from preprocessing workflows

eharkins commented:
Confirmed that the above is our plan for now!

tsibley commented on .github/workflows/preprocess-gisaid.yml:
```diff
@@ -0,0 +1,60 @@
+        required: false
+
+env:
+  S3_DST_BUCKET: ${{ github.event.inputs.trial_name && format('nextstrain-ncov-private/trial-preprocess/{0}', github.event.inputs.trial_name) || 'nextstrain-ncov-private' }}
```
Why uploading to `trial-preprocess/…` instead of `trial/…` like the existing trial build destinations? Seems cleaner to keep the namespacing the same, particularly because it seems like it'd be helpful to optionally start some trial builds from the outputs of a trial preprocessing run of the same name ([rough, incomplete sketch of what I mean](https://gist.github.com/tsibley/e79bb12714acfbaecc7f3768243c7fac)).

tsibley commented on workflow/snakemake_rules/common.smk:
```diff
@@ -170,4 +171,11 @@ def _get_upload_inputs(wildcards):
+    if config.get("upload_preprocessing_files", False):
+        return preprocessing_files
+    elif "filtered" in config["inputs"][origin]:
+        return build_files
```
:raised_hands:
